internal/controller/dnsrecord_healthchecks.go

```diff
@@ -0,0 +1,243 @@
+	} else {
+		log.Log.V(3).Info("reconciling health checks")
+		results, err = r.reconcileHealthChecks(ctx, dnsRecord, config)
+		if err != nil {
+			return err
+		}
+	}
+
+	result := r.reconcileHealthCheckStatus(results, dnsRecord)
+	log.Log.Info("health check status reconcile", "result", result)
+	return result
+}
+
+func (r *DNSRecordReconciler) reconcileHealthChecks(ctx context.Context, dnsRecord *v1alpha1.DNSRecord, config *healthChecksConfig) ([]provider.HealthCheckResult, error) {
+	dnsProvider, err := r.getDNSProvider(ctx, dnsRecord)
+	if err != nil {
+		return nil, err
+	}
+	healthCheckReconciler := dnsProvider.HealthCheckReconciler()
+
+	var results []provider.HealthCheckResult
+	for _, dnsEndpoint := range dnsRecord.Spec.Endpoints {
+		log.Log.Info("reconciling DNS Record health checks", "endpoint", dnsEndpoint)
+		address, ok := provider.GetEndpointAddress(dnsEndpoint)
+		if !ok {
+			log.Log.Info("address not ok", "address", address)
+			continue
+		} else if !net.IsIPv4String(address) {
+			log.Log.Info("address is not an IP", "address", address)
+			continue
+		}
+
+		endpointId, err := idForEndpoint(dnsRecord, dnsEndpoint)
+		if err != nil {
+			return nil, err
+		}
+
+		log.Log.Info("got id", "endpointID", endpointId)
+
+		spec := provider.HealthCheckSpec{
+			Id:               endpointId,
+			Name:             fmt.Sprintf("%s-%s", *dnsRecord.Spec.RootHost, dnsEndpoint.DNSName),
+			Host:             dnsRecord.Spec.RootHost,
+			Path:             config.Endpoint,
+			Port:             config.Port,
+			Protocol:         config.Protocol,
+			FailureThreshold: config.FailureThreshold,
+		}
+
+		probeStatus := r.getProbeStatus(address, dnsRecord)
+
+		result, err := healthCheckReconciler.Reconcile(ctx, spec, dnsEndpoint, probeStatus)
+		log.Log.Info("endpoint provider flags", "flags", dnsEndpoint.ProviderSpecific)
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, result)
+	}
+
+	log.Log.Info("returning health checks", "count", len(results))
+	return results, nil
+}
+
+func (r *DNSRecordReconciler) getProbeStatus(address string, dnsRecord *v1alpha1.DNSRecord) *v1alpha1.HealthCheckStatusProbe {
+	if dnsRecord.Status.HealthCheck == nil || dnsRecord.Status.HealthCheck.Probes == nil {
+		log.Log.Info("status or probes are nil", "status", dnsRecord.Status.HealthCheck)
+		return nil
+	}
+	for _, probeStatus := range dnsRecord.Status.HealthCheck.Probes {
+		if probeStatus.IPAddress == address {
+			return &probeStatus
+		}
+		log.Log.Info("not a match", "ip", probeStatus.IPAddress, "address", address)
+	}
+
+	return nil
+}
+
+func (r *DNSRecordReconciler) reconcileHealthCheckStatus(results []provider.HealthCheckResult, dnsRecord *v1alpha1.DNSRecord) error {
+	probesCondition := metav1.Condition{
+		ObservedGeneration: dnsRecord.Generation,
+		Status:             metav1.ConditionTrue,
+		Reason:             "AllProbesSynced",
+		LastTransitionTime: metav1.Now(),
+		Message:            "all probes synced successfully",
```

Comment:
what if this was fmt.Sprintf("all %v probes synced successfully",len(results))

then lower down you prob don't need the if (len(results) > 0 ) ?